runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1825,6 +1832,106 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+            //      J9Class from the java.lang.Class by way of <classFromJavaLangClass>
+            //
+
+            TR::SymbolReference *symRef = classChild->getOpCode().hasSymbolReference() ? classChild->getSymbolReference() : NULL;
+            TR::SymbolReference *jlcFromClassSymRef = comp()->getSymRefTab()->findOrCreateJavaLangClassFromClassSymbolRef();
+            TR::Node *classOperand = NULL;
+
+            if ((symRef != NULL) && (symRef == jlcFromClassSymRef))
+               {
+               classOperand = classChild->getFirstChild();
+               }
+            else
+               {
+               classOperand = TR::Node::createWithSymRef(TR::aloadi, 1, 1, classChild,
+                                 comp()->getSymRefTab()->findOrCreateClassFromJavaLangClassSymbolRef());
+               }
+
+            TR::Node *loadComponentTypeNode = comp()->fej9()->loadArrayClassComponentType(classOperand);
+            TR::Node *jlcOfComponentTypeNode = NULL;
+
+            if ((classChildConstraint == NULL)
+                || (classChildConstraint->getClassType() == NULL)
+                || !comp()->fej9()->isClassArray(classChildConstraint->getClass()))
+               {
+               // Generate IL of the following form, inlining Class.getComponentType()
+               // without control flow:
+               //
+               //    n99n aselect
+               //    n98n   icmpne
+               //    n97n     iand
+               //    n96n       l2i
+               //    n95n         lloadi <classDepthAndFlags>
+               //    n94n           aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n93n       iconst 0x10000
+               //    n92n     iconst 0
+               //    n91n   aloadi <javaLangClassFromClass>
+               //    n90n     aselect
+               //    n98n       ==> icmpne
+               //    n89n     aloadi <componentClass>
+               //    n94n       ==> aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n94n     ==> aloadi <vft> (or aloadi <classFromJavaLangClass>)
+               //    n88n   aconst NULL
+               //
+               // Note that the same condition testing whether the class is an array
+               // class applies to both aselect operations.  If it is an array, the
+               // <componentClass> field of the j9ArrayClass is loaded, and then the
+               // corresponding java/lang/Class is returned; if it is not an array, a
+               // null reference results.
+               //
+               // Note further that both operands of an aselect are always evaluated,
+               // so if the class is not an array class, whatever bits appear
+               // at the offset of <componentClass> will be loaded for the second
+               // operand of the inner aselect (node n89n, in the example above), but
```

Comment:
Thanks for the detailed comment, it helps understand what is going on with this tree. Is it guaranteed that the offset of `<componentClass>` is valid to access ?

Reply:
Is the IL opcode `aselect` uncollected or collected ? One of the `aselect`  nodes appears to work with collected values (i.e. on the Java heap) whereas the other one works with uncollected ones (J9Classes). I am wondering if by some chance, these `aselect` nodes were commoned across a GC point, then could there be an issue in either case ?

Reply:
> Is it guaranteed that the offset of <componentClass> is valid to access (i.e. won't cause a seg fault) even when the class is not an array class ?

It is the case today, but I could add an additional test &mdash; or an assertion &mdash; that would prevent the transformation from happening if the offset+size of `<componentClass>` would bring us past the end of `j9Class`.

Reply:
Thanks, an assert would future-proof it, so I would suggest adding.

Reply:
No that addresses my concern for now. It seems to be smarter than I was expecting, which is good enough to avoid issues in your sequence (as you checked). Thanks Henry for checking into that.

Reply:
Added a build-time assertion to verify that accessing the `componentType` field of `J9ArrayClass` will not result in an access beyond the end of something that turns out to be a `J9Class` for something that is not an array class in commit df1f93b5301b21f2679f9c544e3ad0fefe23f81c.

I can squash that commit if you are OK with the change.